### PR TITLE
93 new intro quiz

### LINKED ISSUE #93: [3] Have a intro quiz at the landing page
Blocked by #53 

GIVEN I'm a new user
AND at the landing page
WHEN I click on a test your skill button
THEN I can take a quiz to assess my current level

A page for the article with a quiz. Which can mimic a section.

### PR DESCRIPTION
**Pull request by jamiesonreinhard**
Closes #93 

Reviewed mentor branch and implemented changes with several commits on this branch.

Next step is to render a quiz report upon submitting the intro-quiz on the homepage. I'm trying to wrap my head around this logic. 

@caioertai -- will I need to create a new Guest::Sessions controller for this? See attached routing error.
Screen Shot 2021-02-02 at 8 28 39 AM

### CODE REVIEW

**Comment by caioertai** [comment]
Yes, @jamiesonreinhard. I stopped at the router level, but it needs the controller. 

It's the same old flow of Router, Controller, View.

**Comment by caioertai** [comment]
The job of this controller action will be to take the parameters from the form (the users answers to the quiz) and initialize the session with them in order for the report to be rendered.

You can check this controller for reference: `Course::Quiz::SessionsController`.

**Comment by jamiesonreinhard** [comment]
> The job of this controller action will be to take the parameters from the form (the users answers to the quiz) and initialize the session with them in order for the report to be rendered.
> 
> You can check this controller for reference: `Course::Quiz::SessionsController`.

I think I'm on the right track here but maybe you can give me a hint... not sure how to intantiate an instance of a quiz properly here in the highlighted line.

Screen Shot 2021-02-02 at 10 37 07 AM

**Comment by caioertai** [comment]
So. 2 things:
1. You should use `#build` from the quiz builder, for consistency. `#new_quiz` should have been a private method. My bad there. 
2. Once the quiz builder returns you a quiz instance, you should get a quiz session from using `Quiz#session`. No need to instantiate it yourself.

**Comment by caioertai** [comment]
Ignore my second point. I just realized this only works for a new session. It might be a good idea to include the parameters on the method to make it better in the future

**Comment by jamiesonreinhard** [comment]
Got it. So now the quiz builder "build" method is breaking because it isn't receiving a "quizable" with params. The "quizable" should be the article in this case, right? So how can I pass the article from the homepage view, to the guest session controller, to the quiz builder?

**Comment by caioertai** [comment]
Alright. Let's rewind.

Here's the `Course::Quiz::SessionsController#create` 
```ruby
  def create
    @quiz = Quiz.find(params[:quiz_id])
    @quiz_session = Quiz::Session.new(quiz_session_params.merge(quiz: @quiz))

    authorize @quiz_session

    if @quiz_session.save
      redirect_to course_section_path(@quiz.quizable, anchor: "quiz")
    else
      @section = @quiz_session.quiz.quizable
      render "course/sections/show"
    end
  end
```

It needs an instance of a quiz, then it can pass it, along with the quiz session params to the session.
For the the course action the quiz can be found via id, since it's persisted. In case of the guest section, we're going to need to build it again, and the way to build it is by using the `QuizBuilder`. You have the right idea using it, but you're not using its interface properly. Check the initialize:
image

It needs a user, which is a current_user (a guest in this case), and a quizable (the article).

Indeed getting the article in the create is not possible right now since we don't have the article being sent as an argument or present in the body of the POST. We have 2 ways for solving this (3 with a hack):
1. Add the `resources :article` as part of the post route (and change the form accordingly to include `quiz.quizable`). So the form build would become: `[:guest, quiz_session.quiz.quizable, quiz_session]`.
2. Do not change the form in any way and just get the `@quizable` in the controller as `Article.first`.
3. Add a `f.input :article_id, as: :hidden, value: quiz_session.quiz.quizable.id` to the form.

**Comment by jamiesonreinhard** [comment]
Thanks for the explanation there. That makes sense now. I think I'm getting closer, let me know if this looks correct. 

`
class Guest::SessionsController < ApplicationController
  def create
    @quizable = Article.last
    @quiz = QuizBuilder.new(user: current_user, quizable: @quizable).build
    @quiz_session = Quiz::Session.new(quiz_session_params.merge(quiz: @quiz))
    
    if @quiz_session.save
      redirect_to root_path(anchor: "intro-quiz")
    end
  end

  private

  def quiz_session_params
    params.require(:quiz_session).permit(quiz_questions_attributes: [:id, :answer_id])
  end
end
`

This is the new error related to not finding a quiz question: 

Screen Shot 2021-02-02 at 1 10 55 PM

**Comment by caioertai** [comment]
Yeah.... this one might be tricky. Can we have a pair session tomorrow about it, @jamiesonreinhard? I'm concerned and involves some other things I wasn't planning on.

**Comment by caioertai** [comment]
I don't have a single meeting tomorrow, so any time should be fine.

**Comment by jamiesonreinhard** [comment]
Ok great, how about 11am CT?

**Comment by caioertai** [comment]
> Ok great, how about 11am CT?

Perfect.

**Comment by jamiesonreinhard** [comment]
@caioertai I worked a little bit on styling the intro quiz report show page, as well as the actual quiz on the homepage. I made the quiz expandable/collapsible which I think looks okay - I think it could use more styling in the future but works well for now. Let me know what you think.

**Comment by jamiesonreinhard** [comment]
@caioertai also, there is one bug. When you're on the show page for the quiz report and you try to refresh, it breaks. Thoughts on a workaround for that?
Screen Shot 2021-02-09 at 3 47 11 PM

**Comment by caioertai** [comment]
> @caioertai also, there is one bug. When you're on the show page for the quiz report and you try to refresh, it breaks. Thoughts on a workaround for that?
> Screen Shot 2021-02-09 at 3 47 11 PM

Well... we can leave this for another time. We could save the quiz results in the user session (cookies), so if the user refreshes we check for a saved quiz session in the session (yeah... sounds weird, I know)...

**Comment by caioertai** [comment]
Btw. It's not a bug. There IS no quiz session to show unless it comes straight from the form the user submits. So it only lasts one request. It's like this:
- User answers form, and clicks submit
- A POST request with the `body` having the answers is created.
- Controller processes the request and renders show.
- Effectively, you're looking at the rendered results of a POST request at `/sessions/show`

- Now... if you refresh this page, you're just trying to GET `/sessions/show` which doesn't exist.

So what we could do on a GET is check the user cookies for a quiz session.

File: config/routes.rb [hunk 1]
```diff
@@ -2,6 +2,7 @@
   root                  to: "pages#home"
   get  "/about",        to: "pages#about"
   get  "/testimonials", to: "pages#testimonials"
+  get "/guest/sessions", to: "pages#home"
```
Comment (caioertai) [inline config/routes.rb:5]: I'd keep this in thee guest namespace to keep things neater.

```ruby
namespace :guest do
  # ...
  get "/sessions", to: "pages#home"
end
```
this should work there.

**Pull request merged**